packages/application-shell/src/components/custom-view-shell/custom-view-shell.tsx

```diff
@@ -0,0 +1,132 @@
+import {
+  useCallback,
+  useEffect,
+  useRef,
+  useState,
+  type ReactNode,
+} from 'react';
+import type { ApplicationWindow } from '@commercetools-frontend/constants';
+import type { TAsyncLocaleDataProps } from '@commercetools-frontend/i18n';
+import LoadingSpinner from '@commercetools-uikit/loading-spinner';
+import ApplicationShellProvider from '../application-shell-provider';
+import CustomViewAuthenticatedShell from '../custom-view-authenticated-shell';
+import { CustomViewContextProvider, TCustomView } from '../custom-view-context';
+import { CUSTOM_VIEWS_EVENTS_NAMES } from './constants';
+
+declare let window: ApplicationWindow;
+
+type THostContext = {
+  hostUrl: string;
+  userLocale: string;
+  dataLocale: string;
+  customViewConfig: TCustomView;
+  projectKey?: string;
+};
+
+type THostEventData = {
+  source: string;
+  destination: string;
+  eventName: string;
+  eventData: {
+    context: THostContext;
+  };
+};
+
+type TCustomViewShellProps = {
+  customViewId: string;
+  messages: TAsyncLocaleDataProps['applicationMessages'];
+  children: ReactNode;
+};
+
+function CustomViewShell(props: TCustomViewShellProps) {
+  const [hostContext, setHostContext] = useState<THostContext>();
+  const iFrameCommunicationPort = useRef<MessagePort>();
+
+  const hostMessageHandler = useCallback(
+    (event: MessageEvent<THostEventData>) => {
+      if (
```

Comment:
Here is where we process private messages received from the host.
Messages here can only come from the host MC native application.